File: mpagenomics_wrappers/segmentFracB.py
import os
import sys
import subprocess
import zipfile
import getopt


def main(argv):
       
    try:
        opts, args = getopt.getopt(argv,"hc:i:o:f:s:og:fig:t:p:l:u:m:",["chrom=","input=","output=","new_file_path=","settings_type=","output_graph=","zip_figures=","settings_tumor=","outputlog=","log=","userid=","method="])
    except getopt.GetoptError as err:
        print str(err)
        sys.exit(2)
    for opt, arg in opts:
        if opt == '-h':
            print 'extractCNopts.py'
            sys.exit()
        elif opt in ("-c", "--chrom"):
            chromosome = arg
        elif opt in ("-i", "--input"):
            input_file = arg
        elif opt in ("-o", "--output"):
            output_file = arg
        elif opt in ("-f", "--new_file_path"):
            tmp_dir = arg
        elif opt in ("-s", "--settings_type"):
            input_type = arg
        elif opt in ("-og", "--output_graph"):
            output_graph = arg
        elif opt in ("-fig", "--zip_figures"):
            zip_file = arg
        elif opt in ("-t", "--settings_tumor"):
            tumorcsv = arg
        elif opt in ("-p", "--outputlog"):
            outputlog = arg
        elif opt in ("-l", "--log"):
            log = arg
        elif opt in ("-u", "--userid"):
            user_id = arg
        elif opt in ("-m", "--method"):
            method = arg
  
    script_dir=os.path.dirname(os.path.abspath(__file__))
    
    iFile=open(input_file,'r')
    dataSetLine=iFile.readline()
    dataset=dataSetLine.split("\t")[1]
    iFile.close()
      

    if input_type=="dataset":
        input_type=dataset

    if (outputlog=="TRUE"):
        errfile=open(log,'w')
    else:
        errfile=open(os.path.join(tmp_dir,"errfile.log"),'w')
    
    fig_dir=os.path.join("mpagenomics",user_id,"figures",dataset,"segmentation/fracB")

    abs_fig_dir=os.path.join(tmp_dir,fig_dir)
    if (os.path.isdir(abs_fig_dir)) and (output_graph=="TRUE"):
        old_files=os.listdir(abs_fig_dir)
        for ifile in old_files:
            os.remove(os.path.join(abs_fig_dir,ifile))
           
        
    retcode=subprocess.call(["Rscript", os.path.join(script_dir,"segmentFracB.R"),  chromosome, dataset, output_file, tmp_dir, input_type, output_graph, tumorcsv, user_id, method], stdout = errfile, stderr = errfile)
    
    errfile.close()
     
    if (retcode == 0):
        if (os.path.isdir(abs_fig_dir)) and (output_graph=="TRUE"):
        
            new_files=os.listdir(abs_fig_dir)
            zipbuf = zipfile.ZipFile(os.path.join(abs_fig_dir,zip_file), 'w', zipfile.ZIP_DEFLATED)
            for current_file in new_files:
                fn = os.path.join(abs_fig_dir,current_file)
                relfn=fn[len(abs_fig_dir)+len(os.sep):]
                zipbuf.write(fn,relfn)
        sys.exit(retcode)
    else:
        sys.exit(retcode)

if __name__ == "__main__":
    main(main(sys.argv[1:]))

File: mpagenomics_wrappers/segmentFracB.R
#!/usr/bin/env Rscript
# setup R error handling to go to stderr
options( show.error.messages=F, error = function () { cat( geterrmessage(), file=stderr() ); q( "no", 1, F ) } )

# we need that to not crash galaxy with an UTF8 error on German LC settings.
loc <- Sys.setlocale("LC_MESSAGES", "en_US.UTF-8")

library("optparse")

##### Read options
option_list=list(
		make_option("--chrom",type="character",default=NULL, dest="chrom"),
		make_option("--input",type="character",default=NULL, dest="input"),
		make_option("--output",type="character",default=NULL, dest="output"),
		make_option("--new_file_path",type="character",default=NULL, dest="new_file_path"),
		make_option("--settings_type",type="character",default=NULL, dest="settingsType"),
		make_option("--output_graph",type="character",default=NULL, dest="outputgraph"),
		make_option("--zip_figures",type="character",default=NULL, dest="zipfigures"),
		make_option("--settings_tumor",type="character",default=NULL, dest="settingsTypeTumor"),
		make_option("--outputlog",type="character",default=NULL, dest="outputlog"),
		make_option("--log",type="character",default=NULL, dest="log"),
		make_option("--userid",type="character",default=NULL, dest="userid"),
		make_option("--method",type="character",default=NULL, dest="method")
);

opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);

if(is.null(opt$input)){
	print_help(opt_parser)
	stop("input required.", call.=FALSE)
}

#loading libraries

args<-commandArgs(TRUE)

chrom=opt$chrom
datasetFile=opt$input
output=opt$output
tmp_dir=opt$new_file_path
input=opt$settingsType
outputfigures=type.convert(opt$outputgraph)
tumorcsv=opt$settingsTypeTumor
user=opt$userid
method=opt$method
log=opt$log
outputlog=opt$outputlog
outputgraph=opt$outputgraph
zipfigures=opt$zipfigures

#chrom=opt$chrom
#datasetFile=opt$input
#output=opt$output
#tmp_dir=opt$new_file_path
#nbcall=as.numeric(opt$nbcall)
#settingsType=opt$settingsType
#outputfigures=type.convert(opt$outputgraph)
#snp=type.convert(opt$snp)
#tumorcsv=opt$settingsTypeTumor
#cellularity=as.numeric(opt$cellularity)
#user=opt$userid
#method=opt$method
#log=opt$log
#outputlog=opt$outputlog
#outputgraph=opt$outputgraph
#zipfigures=opt$zipfigures

library(MPAgenomics)
workdir=file.path(tmp_dir, "mpagenomics",user)
setwd(workdir)

if (grepl("all",tolower(chrom)) | chrom=="None") {
	chrom_vec=c(1:25)
} else {
	chrom_tmp <- strsplit(chrom,",")
	chrom_vecstring <-unlist(chrom_tmp)
	chrom_vec <- as.numeric(chrom_vecstring)
}


if (outputlog){
	sinklog <- file(log, open = "wt")
	sink(sinklog ,type = "output")
	sink(sinklog, type = "message")
} 


inputDataset=read.table(file=datasetFile,stringsAsFactors=FALSE)
dataset=inputDataset[1,2]



library(MPAgenomics)
workdir=file.path(tmp_dir, "mpagenomics",user)
setwd(workdir)

if (grepl("all",tolower(chrom)) | chrom=="None") {
	chrom_vec=c(1:25)
} else {
	chrom_tmp <- strsplit(chrom,",")
	chrom_vecstring <-unlist(chrom_tmp)
	chrom_vec <- as.numeric(chrom_vecstring)
}

fig_dir = file.path("mpagenomics", user, "figures", dataset, "segmentation","fracB")
abs_fig_dir = file.path(tmp_dir, fig_dir)

if (outputgraph) {
	if (dir.exists(abs_fig_dir)) {
		system(paste0("rm -r ", abs_fig_dir))
	}
}

if (input == 'dataset') {
	segcall=segFracBSignal(dataset,chromosome=chrom_vec, normalTumorArray=tumorcsv, savePlot=outputfigures, method=method)
	
} else {
	input_tmp <- strsplit(input,",")
	input_tmp_vecstring <-unlist(input_tmp)
	input_vecstring = sub("^([^.]*).*", "\\1", input_tmp_vecstring) 
	segcall=segFracBSignal(dataset,chromosome=chrom_vec, normalTumorArray=tumorcsv, listOfFiles=input_vecstring, savePlot=outputfigures, method=method)
	
}
write.table(segcall,output,row.names = FALSE, quote=FALSE, sep = "\t")

if (outputgraph) {	
	setwd(abs_fig_dir)
	files2zip <- dir(abs_fig_dir)
	zip(zipfile = "figures.zip", files = files2zip)
	file.rename("figures.zip",zipfigures)
}

if (outputlog){
	sink(type="output")
	sink(type="message")
	close(sinklog)
} 

#sink(output)
#print(format(segcall))
#sink()


